### ISSUE TITLE
Add validator?

### ISSUE THREAD

**Issue by qzmfranklin** [rules-proto-grpc/rules_proto_grpc#16]
I recently came across a protoc plugin that adds validators to protobuf definitions:

https://github.com/envoyproxy/protoc-gen-validate

Would like to bring into the view of the behold.

**Comment by aaliddell** [comment]
Hi,

This plugin appears to need to target the same output directory as the native plugin, so it possibly may not work with the current implementation that calls each plugin separately to its own directory. I'll have a look to see how it behaves.

**Comment by qzmfranklin** [comment]
I tried this plugin.

It generates .validate.go files under a directory named by the `go_package` option.

That means the `proto_plugin` for it has to be written with `output_directory = True` instead of just using `outputs`, for example:
```bzl
proto_plugin(
    name = 'go_validate_plugin',
    visibility = ['//visibility:public'],
    options = ['lang=go'],
    output_directory = True,
    tool = '@com_github_envoyproxy_protoc_gen_validate//:protoc-gen-validate',
    exclusions = [
        'google/protobuf',
        'validate',
    ],
)
```

**Comment by qzmfranklin** [comment]
The way I'd like to gett it to work is by

```bzl
go_grpc_compile(
    name = 'person_grpc_go',
    deps = [':person_proto'],
)

go_validate_compile(
    name = 'person_validate_go',
    deps = [':person_proto'],
)

go_library(
    name = 'person_pb_go',
    srcs = [
        ':person_grpc_go',
        ':person_validate_go',
    ],
    deps = [
        '@com_github_envoyproxy_protoc_gen_validate//validate:go_default_library',
        '@com_github_golang_protobuf//proto:go_default_library',
        '@com_github_golang_protobuf//ptypes:go_default_library',
        '@org_golang_google_grpc//:go_default_library',
    ],
)

proto_library(
    name = 'person_proto',
    srcs = ['person.proto'],
    deps = [
        '@com_github_envoyproxy_protoc_gen_validate//validate:validate_proto',
    ],
)
```

The `go_grpc_compile` rule came from the existing rule in this repo.  The `go_validate_compile` rule is a custom plugin as shown in the comment above. However, `go_library.srcs` cannot take a directory as input, as is produced by the `go_validate_compile` rule due to the use of the `output_directory = True` attribute.

**Comment by aaliddell** [comment]
Ok, so the new rules this would introduce are:

- `go_validate_compile`
- `gogo_validate_compile`
- `cc_validate_compile`/`cpp_validate_compile`
- `java_validate_compile`
- `py_validate_compile`/`python_validate_compile`

Although the README says the gogo and cpp rules are either experimental or partially complete. To check, is the above pseudocode or something that has been tested?

The `*_library` rules may also be desirable, as they abstract away having to specify the deps every time. The python dependencies would also need updating to import this library's dependencies. Test workspaces or use in the routeguide tests would be useful checks for everything working.

> However, go_library.srcs cannot take a directory as input, as is produced by the go_validate_compile rule due to the use of the output_directory = True attribute.

This may have to be addressed upstream in rules_go beforehand then.

**Comment by qzmfranklin** [comment]
The above pseudocode was modified from codes that actually ran in a canary testing environment.

I only tested with the `protoc-gen-validate` plugin with `lang=go` option. The bazel part was done using code similar (slightly modified due to custom renaming and setup) from the snippet posted above.

I have put together a `go_proto_compile` rule that invokes the `protoc-gen-go` plugin with `plugin=grpc` and `protoc-gen-validate` with `lang=go`. The generated source files are then fed to a `go_library` (as imported from rules_go). The resulting code works nicely together, including the original go proto structs/methods and the `Validate()` methods from the `protoc-gen-validate` plugin.

The first catch is that the {proto}.validate.pb.go file generated by `protoc-gen-validate` is located in a directory named after the `go_package` option in the `.proto` file. This makes it imperative to use `output_directory = True` in the `proto_plugin`. Agree with you that the changes had to be made upstream.

The second catch is that transitive deps (such as google/protobuf/timestamp.proto or any other of my custom protos) are also visited and the generated .validate.pb.go files are also included in the `go_proto_compile` rule. I had to write a small rule to filter out transitive protos and only include files generated from direct protos. But I feel that I should not have to do this. Need advices here.

**Comment by qzmfranklin** [comment]
I did not test their cpp and py rules. I'd like to use them. Just not motivated to actually test them out yet.

Also, just FYI, to make our team's life a little easier, I patched the source code of `protoc-gen-validate` to generate files into the same directory as the .proto file. This is the same behavior as `protoc-gen-go`. I like this behavior much better because the generated file can be expected without having to parse the proto files.

Maybe we can try proposing to `protoc-gen-validate` to add an option to change the path of the generated files. Would you think it worth doing?

**Comment by qzmfranklin** [comment]
[This issue](https://github.com/envoyproxy/protoc-gen-validate/issues/271) received some good input:

https://github.com/envoyproxy/protoc-gen-validate/issues/271#issuecomment-544645772

I believe this should help unblock this issue.

I have not tried it yet. But will give it a shot myself before long.

**Comment by ash2k** [comment]
@qzmfranklin Hey, did you manage to get it working? I'm looking to do the same so a working example would be of great help!

**Comment by ash2k** [comment]
It works for me: https://gitlab.com/gitlab-org/cluster-integration/gitlab-agent/-/merge_requests/132

Would be nice to have it out of the box :)

**Comment by qzmfranklin** [comment]
@ash2k I got it working already. But the solution is rather in-house. It's only a matter of the amount of work to open source it that I'm feeling hesitant about.

**Comment by ashwin153** [comment]
The approach that I took is to create `pgv_go_grpc_compile`, `pgv_go_grpc_library`, `pgv_go_proto_compile`, `pgv_go_proto_compile` macros that are exactly the same as the corresponding `go_*` macros except that they additionally run the following plugin and take a dependency on `@com_envoyproxy_protoc_gen_validate//validate:go_default_library` in the generated `go_library`. I was wondering if the maintainers of `rules_proto_grpc` were interested in up-streaming this change. If so, I would be happy to work on a pull request.

```python
proto_plugin(
    name = "pgv_go_plugin",
    tool = "@com_envoyproxy_protoc_gen_validate//:protoc-gen-validate",
    exclusions = [
        "google/protobuf",
        "validate",
    ],
    options = [
        "lang=go",
        "paths=source_relative",
    ],
    outputs = [
        "{protopath}.pb.validate.go",
    ],
    visibility = [
        "//visibility:public",
    ],
)
```

**Comment by aaliddell** [comment]
👍

Likely yes, I looked at this for 3.0.0 but didn’t have the time to get it working. For go it seems simple enough since it’s just another go dependency. The plugin also supports C++ and Java and I’m not sure about having a go dependency there but perhaps that’s fine.

**Comment by aaliddell** [comment]
I'm adding some experimental rules to 4.0.0 for Go based on the above. Please let me know if it works as desired.
